File: problems/003-binary-tree-serialization/java/Node.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Node {
    String val;
    Node left, right;

    Node(String value) {
        this(value, null, null);
    }

    Node(String value, Node left) {
        this(value, left, null);
    }

    Node(String value, Node left, Node right) {
        this.val = value;
        this.left = left;
        this.right = right;
    }

    public static String serialize(Node root) {
        return root == null ? "" : root.toString();
    }
    
    public static Node deserialize(String treeString) {

        if(treeString == null || treeString.trim().isEmpty()) {
            return null;
        }

        List<String> parts = new ArrayList<>(Arrays.asList(treeString.trim().split(" ")));
        return buildTree(parts);
    }

    private static Node buildTree(List<String> parts) {
        if(parts.isEmpty()) return null;

        String part = parts.remove(0);

        if(part.equals("|")) {
            return null;
        } else {
            Node newNode = new Node(part);
        
            newNode.left = buildTree(parts);
            newNode.right = buildTree(parts);

            return newNode;
        }
    }

    @Override
    public String toString() {

        String output = val + " ";
        
        if(left != null) {
            output += left.toString();
        } else {
            output += "| ";
        }

        if(right != null) {
            output += right.toString();
        } else {
            output += "| ";
        }
        
        return output;
    }
}

File: problems/003-binary-tree-serialization/csharp/Node.cs
using System.Collections.Generic;

public class Node
{
    private string val;
    private Node? left, right;

    public string Val => val;
    public Node? Left => left;
    public Node? Right => right;

    public Node() { }

    public Node(string val, Node? left = null, Node? right = null)
    {
        this.val = val;
        this.left = left;
        this.right = right;
    }

    public static string? Serialize(Node root)
    {
        return root == null ? "" : root.ToString();
    }

    public static Node? Deserialize(string tree)
    {
        if (tree == null)
            return null;

        List<string> parts = tree.Trim().Split(" ").ToList();

        return BuildTree(parts);
    }

    private static Node? BuildTree(List<string> parts)
    {
        if (parts.Count == 0) return null;

        string part = parts.First();
        parts.Remove(part);

        if (part.Equals("|"))
            return null;

        Node newNode = new(part, BuildTree(parts), BuildTree(parts));

        return newNode;
    }

    public override string ToString()
    {
        string output = val + " ";

        if (left != null)
            output += left.ToString();
        else
            output += "| ";

        if (right != null)
            output += right.ToString();
        else
            output += "| ";

        return output;
    }
}
